File: app/lib/features/sync/domain/sync_models.dart
/// Saved sync criteria (backend `SyncCriteria`).
class SyncCriteria {
  const SyncCriteria({
    required this.id,
    required this.label,
    this.countries = const [],
    this.cities = const [],
    this.professions = const [],
    this.interests = const [],
    this.sex,
    this.ageMin,
    this.ageMax,
    this.isActive = true,
    this.lastMatchCount = 0,
  });

  factory SyncCriteria.fromJson(Map<String, dynamic> json) {
    List<String> l(dynamic v) => v is List ? v.map((e) => e.toString()).toList() : const [];
    return SyncCriteria(
      id: (json['id'] ?? '').toString(),
      label: (json['label'] ?? '').toString(),
      countries: l(json['countries']),
      cities: l(json['cities']),
      professions: l(json['professions']),
      interests: l(json['interests']),
      sex: json['sex'] as String?,
      ageMin: (json['ageMin'] as num?)?.toInt(),
      ageMax: (json['ageMax'] as num?)?.toInt(),
      isActive: json['isActive'] as bool? ?? true,
      lastMatchCount: (json['lastMatchCount'] as num?)?.toInt() ?? 0,
    );
  }

  final String id;
  final String label;
  final List<String> countries;
  final List<String> cities;
  final List<String> professions;
  final List<String> interests;
  final String? sex;
  final int? ageMin;
  final int? ageMax;
  final bool isActive;
  final int lastMatchCount;

  Map<String, dynamic> toPayload() => {
        'label': label,
        'countries': countries,
        'cities': cities,
        'professions': professions,
        'interests': interests,
        if (sex != null) 'sex': sex,
        if (ageMin != null) 'ageMin': ageMin,
        if (ageMax != null) 'ageMax': ageMax,
        'isActive': isActive,
      };
}

/// A member proposed for synchronisation, with the backend dedup hint.
class SyncTarget {
  const SyncTarget({
    required this.memberSbcId,
    this.name,
    this.firstName,
    this.profession,
    this.city,
    this.country,
    this.avatarUrl,
    this.phoneNumber,
    this.alreadySynced = false,
  });

  factory SyncTarget.fromJson(Map<String, dynamic> json) => SyncTarget(
        memberSbcId: (json['memberSbcId'] ?? '').toString(),
        name: json['name'] as String?,
        firstName: json['firstName'] as String?,
        profession: json['profession'] as String?,
        city: json['city'] as String?,
        country: json['country'] as String?,
        avatarUrl: json['avatarUrl'] as String?,
        phoneNumber: json['phoneNumber'] as String?,
        alreadySynced: json['alreadySynced'] as bool? ?? false,
      );

  final String memberSbcId;
  final String? name;
  final String? firstName;
  final String? profession;
  final String? city;
  final String? country;
  final String? avatarUrl;
  final String? phoneNumber;
  final bool alreadySynced;

  String get displayName =>
      [firstName, name].where((e) => e != null && e.isNotEmpty).join(' ').trim();
}

/// Result of starting a sync run.
class SyncRunStart {
  const SyncRunStart({
    required this.syncRunId,
    required this.deviceId,
    required this.matchCount,
    required this.pendingCount,
    required this.items,
  });

  factory SyncRunStart.fromJson(Map<String, dynamic> json) => SyncRunStart(
        syncRunId: (json['syncRunId'] ?? '').toString(),
        deviceId: (json['deviceId'] ?? '').toString(),
        matchCount: (json['matchCount'] as num?)?.toInt() ?? 0,
        pendingCount: (json['pendingCount'] as num?)?.toInt() ?? 0,
        items: (json['items'] as List<dynamic>? ?? [])
            .whereType<Map<String, dynamic>>()
            .map(SyncTarget.fromJson)
            .toList(),
      );

  final String syncRunId;
  final String deviceId;
  final int matchCount;
  final int pendingCount;
  final List<SyncTarget> items;
}

/// "Mes contacts SBC" dashboard counters.
class SyncSummary {
  const SyncSummary({
    required this.syncedCount,
    required this.pendingCount,
    required this.failedCount,
    required this.activeCriteria,
    required this.currentMatches,
    this.lastSyncAt,
  });

  factory SyncSummary.fromJson(Map<String, dynamic> json) => SyncSummary(
        syncedCount: (json['syncedCount'] as num?)?.toInt() ?? 0,
        pendingCount: (json['pendingCount'] as num?)?.toInt() ?? 0,
        failedCount: (json['failedCount'] as num?)?.toInt() ?? 0,
        activeCriteria: (json['activeCriteria'] as num?)?.toInt() ?? 0,
        currentMatches: (json['currentMatches'] as num?)?.toInt() ?? 0,
        lastSyncAt: json['lastSyncAt'] == null
            ? null
            : DateTime.tryParse(json['lastSyncAt'].toString()),
      );

  final int syncedCount;
  final int pendingCount;
  final int failedCount;
  final int activeCriteria;
  final int currentMatches;
  final DateTime? lastSyncAt;
}

/// One member already written (or attempted) to a device — "Mes contacts SBC"
/// (cahier §16). `status` mirrors the backend SyncStatus enum.
class SyncedContact {
  const SyncedContact({
    required this.memberSbcId,
    required this.status,
    this.name,
    this.firstName,
    this.profession,
    this.city,
    this.country,
    this.avatarUrl,
    this.phoneNumber,
    this.syncedAt,
  });

  factory SyncedContact.fromJson(Map<String, dynamic> json) => SyncedContact(
        memberSbcId: (json['memberSbcId'] ?? '').toString(),
        status: (json['status'] ?? 'PENDING').toString(),
        name: json['name'] as String?,
        firstName: json['firstName'] as String?,
        profession: json['profession'] as String?,
        city: json['city'] as String?,
        country: json['country'] as String?,
        avatarUrl: json['avatarUrl'] as String?,
        phoneNumber: json['phoneNumber'] as String?,
        syncedAt: json['syncedAt'] == null
            ? null
            : DateTime.tryParse(json['syncedAt'].toString()),
      );

  final String memberSbcId;
  final String status;
  final String? name;
  final String? firstName;
  final String? profession;
  final String? city;
  final String? country;
  final String? avatarUrl;
  final String? phoneNumber;
  final DateTime? syncedAt;

  String get displayName {
    final n = [firstName, name].where((e) => e != null && e.isNotEmpty).join(' ').trim();
    return n.isEmpty ? 'Membre SBC' : n;
  }

  String get location =>
      [city, country].whereType<String>().where((e) => e.isNotEmpty).join(', ');
}

/// One synchronisation run, for the history view (cahier §17).
class SyncRunEntry {
  const SyncRunEntry({
    required this.id,
    required this.status,
    required this.matchCount,
    required this.syncedCount,
    required this.failedCount,
    this.error,
    this.startedAt,
    this.finishedAt,
  });

  factory SyncRunEntry.fromJson(Map<String, dynamic> json) => SyncRunEntry(
        id: (json['id'] ?? '').toString(),
        status: (json['status'] ?? 'PENDING').toString(),
        matchCount: (json['matchCount'] as num?)?.toInt() ?? 0,
        syncedCount: (json['syncedCount'] as num?)?.toInt() ?? 0,
        failedCount: (json['failedCount'] as num?)?.toInt() ?? 0,
        error: json['error'] as String?,
        startedAt: json['startedAt'] == null
            ? null
            : DateTime.tryParse(json['startedAt'].toString()),
        finishedAt: json['finishedAt'] == null
            ? null
            : DateTime.tryParse(json['finishedAt'].toString()),
      );

  final String id;
  final String status;
  final int matchCount;
  final int syncedCount;
  final int failedCount;
  final String? error;
  final DateTime? startedAt;
  final DateTime? finishedAt;
}

File: ios/SBCContacts/Features/Sync/SyncModels.swift
import Foundation

/// Saved sync criteria (backend `SyncCriteria`).
struct SyncCriteria: Identifiable, Sendable, Equatable {
    var id: String
    var label: String
    var countries: [String] = []
    var cities: [String] = []
    var professions: [String] = []
    var interests: [String] = []
    var sex: String?
    var ageMin: Int?
    var ageMax: Int?
    var isActive = true
    var lastMatchCount = 0

    var payload: CriteriaPayload {
        CriteriaPayload(
            label: label, countries: countries, cities: cities, professions: professions,
            interests: interests, sex: sex, ageMin: ageMin, ageMax: ageMax, isActive: isActive
        )
    }
}

extension SyncCriteria: Decodable {
    private enum CodingKeys: String, CodingKey {
        case id, label, countries, cities, professions, interests, sex, ageMin, ageMax, isActive, lastMatchCount
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.string(.id) ?? ""
        label = c.string(.label) ?? ""
        countries = c.strings(.countries)
        cities = c.strings(.cities)
        professions = c.strings(.professions)
        interests = c.strings(.interests)
        sex = c.string(.sex)
        ageMin = c.int(.ageMin)
        ageMax = c.int(.ageMax)
        isActive = c.bool(.isActive) ?? true
        lastMatchCount = c.int(.lastMatchCount) ?? 0
    }
}

/// One person who saved YOU to their phone (cahier §21). Only confirmed device
/// writes reach this list, so "X t'a enregistré" is a fact, not an inference.
struct SavedMeEntry: Identifiable, Sendable, Equatable {
    var actorSbcId: String
    var savedAt: Date
    var name: String?
    var profession: String?
    var city: String?
    var country: String?
    var avatarUrl: String?
    var phoneNumber: String?

    /// Whether you have already saved them back.
    var alreadySaved = false

    var id: String { actorSbcId }

    var displayName: String {
        let n = (name ?? "").trimmingCharacters(in: .whitespaces)
        return n.isEmpty ? "Membre SBC" : n
    }

    var location: String {
        [city, country].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: ", ")
    }
}

extension SavedMeEntry: Decodable {
    private enum CodingKeys: String, CodingKey {
        case actorSbcId, savedAt, name, profession, city, country, avatarUrl, phoneNumber
        case alreadySaved
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        actorSbcId = c.string(.actorSbcId) ?? ""
        savedAt = c.date(.savedAt) ?? Date()
        name = c.string(.name)
        profession = c.string(.profession)
        city = c.string(.city)
        country = c.string(.country)
        avatarUrl = c.string(.avatarUrl)
        phoneNumber = c.string(.phoneNumber)
        alreadySaved = c.bool(.alreadySaved) ?? false
    }
}

/// Body for create / update / ad-hoc preview.
struct CriteriaPayload: Encodable, Sendable, Equatable {
    var label: String
    var countries: [String]
    /// The backend mirrors SBC's `region` into Member.city, so régions go here.
    var cities: [String]
    var professions: [String]
    var interests: [String]
    var sex: String?
    var ageMin: Int?
    var ageMax: Int?
    var isActive: Bool?

    private enum CodingKeys: String, CodingKey {
        case label, countries, cities, professions, interests, sex, ageMin, ageMax, isActive
    }

    /// `ageMin` / `ageMax` are sent as explicit nulls: on an update, leaving the
    /// keys out would keep whatever range the criteria already had.
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(label, forKey: .label)
        try c.encode(countries, forKey: .countries)
        try c.encode(cities, forKey: .cities)
        try c.encode(professions, forKey: .professions)
        try c.encode(interests, forKey: .interests)
        try c.encodeIfPresent(sex, forKey: .sex)
        try c.encode(ageMin, forKey: .ageMin)
        try c.encode(ageMax, forKey: .ageMax)
        try c.encodeIfPresent(isActive, forKey: .isActive)
    }
}

/// A member proposed for synchronisation, with the backend dedup hint.
struct SyncTarget: Sendable, Equatable {
    var memberSbcId: String
    var name: String?
    var firstName: String?
    var profession: String?
    var city: String?
    var country: String?
    var avatarUrl: String?
    var phoneNumber: String?
    var alreadySynced = false

    var displayName: String {
        [firstName, name].compactMap { $0 }.filter { !$0.isEmpty }
            .joined(separator: " ").trimmingCharacters(in: .whitespaces)
    }
}

extension SyncTarget: Decodable {
    private enum CodingKeys: String, CodingKey {
        case memberSbcId, name, firstName, profession, city, country, avatarUrl, phoneNumber, alreadySynced
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        memberSbcId = c.string(.memberSbcId) ?? ""
        name = c.string(.name)
        firstName = c.string(.firstName)
        profession = c.string(.profession)
        city = c.string(.city)
        country = c.string(.country)
        avatarUrl = c.string(.avatarUrl)
        phoneNumber = c.string(.phoneNumber)
        alreadySynced = c.bool(.alreadySynced) ?? false
    }
}

/// Result of starting a sync run.
struct SyncRunStart: Sendable {
    var syncRunId: String
    var deviceId: String
    var matchCount: Int
    var pendingCount: Int
    var items: [SyncTarget]
}

extension SyncRunStart: Decodable {
    private enum CodingKeys: String, CodingKey {
        case syncRunId, deviceId, matchCount, pendingCount, items
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        syncRunId = c.string(.syncRunId) ?? ""
        deviceId = c.string(.deviceId) ?? ""
        matchCount = c.int(.matchCount) ?? 0
        pendingCount = c.int(.pendingCount) ?? 0
        items = ((try? c.decodeIfPresent([Lossy<SyncTarget>].self, forKey: .items)) ?? []).compactMap(\.value)
    }
}

/// One outcome reported back after writing to the phone book.
struct SyncRunResult: Encodable, Sendable, Equatable {
    enum Status: String, Encodable, Sendable {
        case synced = "SYNCED", failed = "FAILED"
    }

    var memberSbcId: String
    var deviceContactId: String?
    var status: Status
}

/// "Mes contacts SBC" dashboard counters.
struct SyncSummary: Sendable, Equatable {
    var syncedCount = 0
    var pendingCount = 0
    var failedCount = 0
    var activeCriteria = 0

    /// Every criterion, active or paused — the denominator of the dashboard ring.
    var criteriaCount = 0
    var currentMatches = 0
    var favoritesCount = 0

    /// How many members have saved your contact (§21).
    var savedMeCount = 0
    var lastSyncAt: Date?
}

extension SyncSummary: Decodable {
    private enum CodingKeys: String, CodingKey {
        case syncedCount, pendingCount, failedCount, activeCriteria, criteriaCount
        case currentMatches, favoritesCount, savedMeCount, lastSyncAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        syncedCount = c.int(.syncedCount) ?? 0
        pendingCount = c.int(.pendingCount) ?? 0
        failedCount = c.int(.failedCount) ?? 0
        activeCriteria = c.int(.activeCriteria) ?? 0
        // Falls back to the active count so an older backend still renders a
        // sane "N / M critères" instead of "3 / 0".
        criteriaCount = c.int(.criteriaCount) ?? activeCriteria
        currentMatches = c.int(.currentMatches) ?? 0
        favoritesCount = c.int(.favoritesCount) ?? 0
        savedMeCount = c.int(.savedMeCount) ?? 0
        lastSyncAt = c.date(.lastSyncAt)
    }
}

/// One member already written (or attempted) to a device — "Mes contacts SBC"
/// (cahier §16). `status` mirrors the backend SyncStatus enum.
struct SyncedContact: Identifiable, Sendable, Equatable {
    var memberSbcId: String
    var status: String
    var name: String?
    var firstName: String?
    var profession: String?
    var city: String?
    var country: String?
    var avatarUrl: String?
    var phoneNumber: String?
    var syncedAt: Date?

    var id: String { memberSbcId }

    var displayName: String {
        let n = [firstName, name].compactMap { $0 }.filter { !$0.isEmpty }
            .joined(separator: " ").trimmingCharacters(in: .whitespaces)
        return n.isEmpty ? "Membre SBC" : n
    }

    var location: String {
        [city, country].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: ", ")
    }
}

extension SyncedContact: Decodable {
    private enum CodingKeys: String, CodingKey {
        case memberSbcId, status, name, firstName, profession, city, country, avatarUrl, phoneNumber, syncedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        memberSbcId = c.string(.memberSbcId) ?? ""
        status = c.string(.status) ?? "PENDING"
        name = c.string(.name)
        firstName = c.string(.firstName)
        profession = c.string(.profession)
        city = c.string(.city)
        country = c.string(.country)
        avatarUrl = c.string(.avatarUrl)
        phoneNumber = c.string(.phoneNumber)
        syncedAt = c.date(.syncedAt)
    }
}

/// One synchronisation run, for the history view (cahier §17).
struct SyncRunEntry: Identifiable, Sendable, Equatable {
    var id: String
    var status: String
    var matchCount: Int
    var syncedCount: Int
    var failedCount: Int
    var error: String?
    var startedAt: Date?
    var finishedAt: Date?
}

extension SyncRunEntry: Decodable {
    private enum CodingKeys: String, CodingKey {
        case id, status, matchCount, syncedCount, failedCount, error, startedAt, finishedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.string(.id) ?? ""
        status = c.string(.status) ?? "PENDING"
        matchCount = c.int(.matchCount) ?? 0
        syncedCount = c.int(.syncedCount) ?? 0
        failedCount = c.int(.failedCount) ?? 0
        error = c.string(.error)
        startedAt = c.date(.startedAt)
        finishedAt = c.date(.finishedAt)
    }
}
